playback/src/audio_backend/pipewire.rs

```diff
@@ -0,0 +1,338 @@
+use super::{Open, Sink, SinkAsBytes, SinkError, SinkResult};
+use crate::config::AudioFormat;
+use crate::convert::Converter;
+use crate::decoder::AudioPacket;
+use crate::{NUM_CHANNELS, SAMPLE_RATE};
+use std::sync::{
+    Arc,
+    atomic::{AtomicBool, Ordering},
+};
+use std::thread;
+use thiserror::Error;
+
+use libspa::sys as spa_sys;
+use pipewire as pw;
+use pw::{properties::properties, spa};
+use ringbuf::{
+    HeapRb,
+    traits::{Consumer, Producer, Split},
+};
+use spa::pod::Pod;
+
+type RingProducer = ringbuf::wrap::caching::Caching<
+    std::sync::Arc<ringbuf::SharedRb<ringbuf::storage::Heap<u8>>>,
+    true,
+    false,
+>;
+type RingConsumer = ringbuf::wrap::caching::Caching<
+    std::sync::Arc<ringbuf::SharedRb<ringbuf::storage::Heap<u8>>>,
+    false,
+    true,
+>;
+
+// Ring buffer size: 1 second of audio at max quality (F64, stereo, 44.1kHz)
+const RING_BUFFER_SIZE: usize = 44100 * 2 * 8;
+
+#[derive(Debug, Error)]
+enum PipeWireError {
+    #[error("<PipeWireSink> Failed to Create Main Loop: {0}")]
+    MainLoopCreation(String),
+
+    #[error("<PipeWireSink> Failed to Create Stream: {0}")]
+    StreamCreation(String),
+
+    #[error("<PipeWireSink> Failed to Connect Stream: {0}")]
+    StreamConnect(String),
+
+    #[error("<PipeWireSink> Stream Not Connected")]
+    NotConnected,
+}
+
+impl From<PipeWireError> for SinkError {
+    fn from(e: PipeWireError) -> SinkError {
+        use PipeWireError::*;
+        let es = e.to_string();
+        match e {
+            MainLoopCreation(_) | StreamCreation(_) | StreamConnect(_) => {
+                SinkError::ConnectionRefused(es)
+            }
+            NotConnected => SinkError::NotConnected(es),
+        }
+    }
+}
+
+pub struct PipeWireSink {
+    format: AudioFormat,
+    // Lock-free ring buffer producer for audio data
+    producer: Option<RingProducer>,
+    // Flag to signal thread to stop
+    quit_flag: Arc<AtomicBool>,
+    initialized: bool,
+    _main_loop_handle: Option<thread::JoinHandle<()>>,
+}
+
+fn calculate_sample_size(format: AudioFormat) -> usize {
+    use AudioFormat::*;
+    match format {
+        F64 => 8,
+        F32 | S32 | S24 => 4,
+        S24_3 => 3,
+        S16 => 2,
+    }
+}
+
+fn convert_audio_format(format: AudioFormat) -> spa::param::audio::AudioFormat {
+    use AudioFormat::*;
+    match format {
+        F64 => spa::param::audio::AudioFormat::F64LE,
+        F32 => spa::param::audio::AudioFormat::F32LE,
+        S32 => spa::param::audio::AudioFormat::S32LE,
+        S24 => spa::param::audio::AudioFormat::S24_32LE,
+        S24_3 => spa::param::audio::AudioFormat::S24LE,
+        S16 => spa::param::audio::AudioFormat::S16LE,
+    }
+}
+
+impl Open for PipeWireSink {
+    fn open(_device: Option<String>, format: AudioFormat) -> Self {
```

Comment:
I think we could *not* ignore `device` and around line 238 where you create the stream add:

```rust
*pw::keys::TARGET_OBJECT => device.as_deref().unwrap_or("")
```

Probably it would then also need:
1. logic to use the "default" device (skipping `TARGET_OBJECT`?)
2. support for object enumeration